x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/rank/textsimilarity/TextSimilarityRankTests.java

```diff
@@ -0,0 +1,180 @@
+                return;
+            }
+
+            if (failInference.get()) {
+                System.out.println("failInference");
+                listener.onFailure(new ElasticsearchException("rerank inference call failed"));
+            } else {
+                // For inference action respond with rerank results
+                ActionResponse response = new InferenceAction.Response(new RankedDocsResults(RANKED_DOCS));
+                listener.onResponse((Response) response);
+            }
+        }
+    }
+
+    private Client client;
+
+    @Override
+    protected Collection<Class<? extends Plugin>> getPlugins() {
+        return List.of(InferencePlugin.class, TestPlugin.class);
+    }
+
+    @Before
+    public void setup() {
+        failInference.set(false);
+
+        // Initialize index with a few documents
+        client = client();
+        for (int i = 0; i < 5; i++) {
+            client.prepareIndex("my-index").setSource(Collections.singletonMap("text", "text " + i)).get();
+        }
+        client.admin().indices().prepareRefresh("my-index").get();
+    }
+
+    public void testRerank() {
+        ElasticsearchAssertions.assertNoFailuresAndResponse(
+            // Execute search with text similarity reranking
+            client.prepareSearch()
+                .setRankBuilder(new TextSimilarityRankBuilder("text", "my-rerank-model", "my query", 100, 0.0f))
+                .setQuery(QueryBuilders.matchAllQuery()),
+            response -> {
+                // Verify order, rank and score of results
+                SearchHit[] hits = response.getHits().getHits();
+                assertEquals(5, hits.length);
+                assertHitHasRankScoreAndText(hits[0], 1, 0.9f, "text 3");
+                assertHitHasRankScoreAndText(hits[1], 2, 0.8f, "text 0");
+                assertHitHasRankScoreAndText(hits[2], 3, 0.7f, "text 2");
+                assertHitHasRankScoreAndText(hits[3], 4, 0.6f, "text 1");
+                assertHitHasRankScoreAndText(hits[4], 5, 0.5f, "text 4");
+            }
+        );
+    }
+
+    public void testRerankWithMinScore() {
+        ElasticsearchAssertions.assertNoFailuresAndResponse(
+            // Execute search with text similarity reranking
+            client.prepareSearch()
+                .setRankBuilder(new TextSimilarityRankBuilder("text", "my-rerank-model", "my query", 100, 0.7f))
+                .setQuery(QueryBuilders.matchAllQuery()),
+            response -> {
+                // Verify order, rank and score of results
+                SearchHit[] hits = response.getHits().getHits();
+                assertEquals(3, hits.length);
+                assertHitHasRankScoreAndText(hits[0], 1, 0.9f, "text 3");
+                assertHitHasRankScoreAndText(hits[1], 2, 0.8f, "text 0");
+                assertHitHasRankScoreAndText(hits[2], 3, 0.7f, "text 2");
+            }
+        );
+    }
+
+    public void testRerankInferenceFailure() {
```

Comment:
Thank you! Could we have a test now that ensures we throw when we don't get the right number of rankings from the inference service?

Reply:
Added a test case (https://github.com/elastic/elasticsearch/pull/109813/commits/a8564a2e0cdf94e5b19909fbe8479aa707a4dead).
I feel the in-place behavior change for the static `TestFilter` is a bit clunky, but I don't have a better idea as it's initialized once.

Reply:
I think that there is no guarantee of test order execution, so maybe we could be met with a scenario where this runs first, changes the `runMode`, and then the first couple of tests  (e.g. `testRerankWithMinScore`) which assume `TestFilterRunMode.NORMAL`, might fail as a result of this?

Reply:
Another option could be to override the `TextSimilarityRankFeaturePhaseRankCoordinatorContext#generateRequest` and pass any needed metadata/flags along with the `taskSettings` configuration.

Reply:
I did some refactoring: extracted `TextSimilarityTestPlugin` into its own class and added support for a context that returns invalid inference results for that test (https://github.com/elastic/elasticsearch/pull/109813/commits/444bf09c0388698c69abc1b79f303eeecb8d0bf5).
Now `TextSimilarityRankTests` and `TextSimilarityRankMultiNodeTests` both use this plugin.